### PR TITLE
fix boot ssh key setting

### LINKED ISSUE #32: ssh-key setting for rescue mode gives errors
##### SUMMARY
Setting ssh-keys with community.hcloud.boot gives errors.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
community.hrobot.boot

##### ANSIBLE VERSION
`pip-installed ansible 4.9.0 [core 2.11.6] and also for ansible 5.1.0 [core 2.12.1]`

##### COLLECTION VERSION
  1.2.1

##### CONFIGURATION
```
[defaults]
nocows = 1
stdout_callback = yaml

interpreter_python = auto_silent

library = /usr/share/ansible/library


retry_files_enabled = False
force_handlers = True


[ssh_connection]
pipelining = True
```

##### OS / ENVIRONMENT
Debian 11 amd64 

##### STEPS TO REPRODUCE
with playbook:
```
- hosts: all
  remote_user: root
  gather_facts: false
  vars:
    hetzner_rescue_system_ipv6: "{{hetzner_ipv6_subnet|ansible.netcommon.ipmath(2)}}"
  tasks:

    - name: Add an SSH key
      local_action: community.hrobot.ssh_key
      args:
        hetzner_user: "{{hetzner_robot_api_username}}"
        hetzner_password: "{{hetzner_robot_api_password}}"
        state: present
        name: rescue
        public_key: "{{ lookup('file', '~/.ssh/id_ed25519.pub') }}"

    - name: List all SSH keys
      local_action: community.hrobot.ssh_key_info
      args:
        hetzner_user: "{{hetzner_robot_api_username}}"
        hetzner_password: "{{hetzner_robot_api_password}}"
      register: hrobot_ssh_key_info

    - name: Enable a rescue system (64bit Linux) for the next boot
      local_action: community.hrobot.boot
      args:
        server_number: "{{hetzner_server_number}}"
        hetzner_user: "{{hetzner_robot_api_username}}"
        hetzner_password: "{{hetzner_robot_api_password}}"
        rescue:
          os: linux
          authorized_keys: "{{ hrobot_ssh_key_info | json_query('ssh_keys[*].fingerprint') }}"
```

##### EXPECTED RESULTS

```
TASK [Enable a rescue system (64bit Linux) for the next boot] *********************************************************************************************************************************
ok: [s1 -> localhost] => changed=false 
  configuration_type: rescue
  password: xxx
changed: [s2 -> localhost] => changed=true 
  configuration_type: rescue
  password: xxx
changed: [s3 -> localhost] => changed=true 
  configuration_type: rescue
  password: xxx
```

##### ACTUAL RESULTS

```
TASK [Enable a rescue system (64bit Linux) for the next boot] *********************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: '<' not supported between instances of 'dict' and 'dict'
fatal: [s1 -> localhost]: FAILED! => changed=false 
  module_stderr: |-
    Traceback (most recent call last):
      File "/home/ka/.ansible/tmp/ansible-tmp-1640722028.009722-72412-92687179374263/AnsiballZ_boot.py", line 100, in <module>
        _ansiballz_main()
      File "/home/ka/.ansible/tmp/ansible-tmp-1640722028.009722-72412-92687179374263/AnsiballZ_boot.py", line 92, in _ansiballz_main
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
      File "/home/ka/.ansible/tmp/ansible-tmp-1640722028.009722-72412-92687179374263/AnsiballZ_boot.py", line 40, in invoke_module
        runpy.run_module(mod_name='ansible_collections.community.hrobot.plugins.modules.boot', init_globals=dict(_module_fqn='ansible_collections.community.hrobot.plugins.modules.boot', _modlib_path=modlib_path),
      File "/usr/lib/python3.9/runpy.py", line 210, in run_module
        return _run_module_code(code, init_globals, run_name, mod_spec)
      File "/usr/lib/python3.9/runpy.py", line 97, in _run_module_code
        _run_code(code, mod_globals, init_globals,
      File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "/tmp/ansible_community.hrobot.boot_payload_oaq5ye2n/ansible_community.hrobot.boot_payload.zip/ansible_collections/community/hrobot/plugins/modules/boot.py", line 430, in <module>
      File "/tmp/ansible_community.hrobot.boot_payload_oaq5ye2n/ansible_community.hrobot.boot_payload.zip/ansible_collections/community/hrobot/plugins/modules/boot.py", line 397, in main
    TypeError: '<' not supported between instances of 'dict' and 'dict'
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1
fatal: [s2 -> localhost]: FAILED! => changed=false 
  msg: 'Request failed: 400 INVALID_INPUT (invalid input)'
fatal: [s3 -> localhost]: FAILED! => changed=false 
  msg: 'Request failed: 400 INVALID_INPUT (invalid input)'

```

### PR DESCRIPTION
**Pull request by chaotika**
##### SUMMARY
Fixes #32

##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
community.hrobot.boot

##### ADDITIONAL INFORMATION
- changes encoding of authorized_key for lists
- fixes idempotence check for ssh keys

### CODE REVIEW

File: plugins/modules/boot.py [hunk 1]
```diff
@@ -384,15 +384,22 @@ def main():
                 option = module.params[option_name][option_key]
                 if option is None or option == []:
                     continue
-                data[data_key] = option
+
+                if isinstance(option, list):
+                    data[data_key+'[]'] = option
```
Comment (felixfontein) [inline plugins/modules/boot.py:389]: What are the `[]` for? The API docs do not mention that.

File: plugins/modules/boot.py [hunk 2]
```diff
@@ -384,15 +384,22 @@ def main():
...
                 needs_change = False
                 for option_key, (result_key, data_key) in options.items():
                     should = module.params[option_name][option_key]
                     if should is None:
                         continue
-                    has = existing.get(data_key)
+                    if option_key == 'authorized_keys':
+                        has = list(map(lambda x: x['key']['fingerprint'], existing.get(data_key)))
+                    else:
+                        has = existing.get(data_key)
```
Comment (felixfontein) [inline plugins/modules/boot.py:402]:
```suggestion
                    has = existing.get(data_key)
                    if option_key == 'authorized_keys':
                        has = [x['key']['fingerprint'] for x in has]
```
(It's too bad that the API docs do not provide any example or schema for how `authorized_key` looks like...)

File: plugins/modules/boot.py [hunk 3]
```diff
@@ -415,7 +422,7 @@ def main():
                     result, dummy = fetch_url_json(
                         module,
                         url,
-                        data=urlencode(data),
+                        data=urlencode(data,True),
```
Comment (felixfontein) [inline plugins/modules/boot.py:425]:
```suggestion
                        data=urlencode(data, True),
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

**Comment by chaotika** [comment]
> Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to)?

I gave it a try. Is this sufficient?

File: changelogs/fragments/33-fix-boot-ssh-key-setting.yaml [hunk 4]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - bugfix for incorrect handling of ssh-keys for the community.hcloud.boot module. Trying to set the boot record with the module lets the API return errors (reported in https://github.com/ansible-collections/community.hrobot/issues/32). This is and the idempotence for ssh-keys is fixed with the https://github.com/ansible-collections/community.hrobot/pull/33 MR.
```
Comment (felixfontein) [inline changelogs/fragments/33-fix-boot-ssh-key-setting.yaml:2]: Following the exact format explained [here](https://docs.ansible.com/ansible/devel/community/development_process.html#changelog-fragment-entry-format):
```suggestion
  - boot - fix incorrect handling of ssh-keys (https://github.com/ansible-collections/community.hrobot/issues/32, https://github.com/ansible-collections/community.hrobot/pull/33).
```

File: plugins/modules/boot.py (hunk 1 above)
Comment (felixfontein) [reply plugins/modules/boot.py:389]: I tried this out, and cannot figure out why you add the `[]`. It does work fine without, and that's also what the API docs document. Please remove this.

File: plugins/modules/boot.py (hunk 2 above)
Comment (felixfontein) [reply plugins/modules/boot.py:403]: For the record: the list entries are of the form `{'key': {'fingerprint': 'aa:bb:cc:...', 'name': 'foo', 'size': 4096, 'type': 'RSA'}}`.

Comment (felixfontein) [reply plugins/modules/boot.py:403]: (This also needs to be adjusted in the tests. They right now assume that this is a list of fingerprints.)

File: plugins/modules/boot.py [hunk 5]
```diff
@@ -384,15 +384,23 @@ def main():
                 option = module.params[option_name][option_key]
                 if option is None or option == []:
                     continue
-                data[data_key] = option
+
+                if isinstance(option, list):
+                    data[data_key+'[]'] = option
+                else:
+                    data[data_key] = option
```
Comment (felixfontein) [inline plugins/modules/boot.py:391]:
```suggestion
                data[data_key] = option
```

File: changelogs/fragments/33-fix-boot-ssh-key-setting.yaml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - boot - fix incorrect handling of ssh-keys (https://github.com/ansible-collections/community.hrobot/issues/32, https://github.com/ansible-collections/community.hrobot/pull/33).
```
Comment (felixfontein) [inline changelogs/fragments/33-fix-boot-ssh-key-setting.yaml:2]:
```suggestion
  - boot - fix incorrect handling of SSH authorized keys (https://github.com/ansible-collections/community.hrobot/issues/32, https://github.com/ansible-collections/community.hrobot/pull/33).
```

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@chaotika thanks for fixing this!

**Pull request merged**